### DISCUSSION TITLE
`new RegExp` vs regex literal

### DISCUSSION THREAD

**Discussion by ohmree** [tree-sitter/tree-sitter#1815] [Q&A] [answered]
When parsing the contents of escape sequences that are used for inputting characters by hex value like `\xhh`, `\uhhhh` and `\Uhhhhhhhh` there is a tiny bit of repetition that can be extracted into a function (although with the `Hex_Digit` unicode property it gets way more self-descriptive and terse).
Here's what I do now:
```js
_shared_escape_sequence_content: $ =>
  choice(/[\\'"bfnrt]/, seq('x', /\p{Hex_Digit}{2}/u)),
_string_escape_sequence_content: $ =>
  choice(
    $._shared_escape_sequence_content,
    '\n',
    seq('u', /\p{Hex_Digit}{4}/u),
    seq('U', /\p{Hex_Digit}{8}/u),
  ),
```
And here's the common "N hex digits" pattern extracted into a function:
```js
/**
 * @param {number} length
 */
function hexNumber(length) {
  // NOTE: without the `u' flag we'd have to do the following,
  // which definitely justifies the existence of this function:
  // return new RegExp(`[0-9a-fA-F]{${length}}`);
  // NOTE: this is not nearly as bad when used as a literal but is still nicer in a function:
  return new RegExp(`\\p{Hex_Digit}{${length}}`, 'u');
}
```
In regular JS I'd either just use a literal here (with the `u` flag and the `Hex_Digit` property) or store the dynamically-generated regexes in variables to avoid redundant recompilation.
Is this a concern in tree-sitter grammars too, since they use a custom regex engine?

P.S. how does the `RegExp` constructor even work in tree-sitter if it doesn't use node's regex engine?
I can see how one could transform a regex literal to c code at compile time, but with the constructor I can't see a way around having to evaluate the arguments and any of their dependencies.
Since the grammar code can't directly do runtime I/O (or can it?) I'd imagine everything should be statically analyzable, making functions behave like rust's `const fn`s.
Assuming the assumptions made previously are correct I can see how in theory one could call into node to evaluate the pattern and flags arguments then feed the resulting strings into the custom regex engine just like with regex literals.
Is this similar to what tree-sitter does or am I way off the mark?

**Comment by SamB** [comment, marked as answer]
It seems that using JavaScript regexes to represent regexes is just a cute trick tree-sitter does, mostly because of the literal syntax. It could have just used a custom type otherwise.

**Comment by sogaiu** [comment]
[Possibly of interest](https://github.com/sogaiu/ts-questions/blob/6e0aac7ac565023bdc571db865c2ea582d6560d7/questions/what-regex-features-are-supported/README.md) -- corrections, suggestions, etc. welcome.
